File: Pragma_Examples/OpenMP/Fortran/3_vecadd/0_vecadd_portyourself/vecadd.f90
! Copyright (c) 2022 Advanced Micro Devices, Inc. All rights reserved.
! 
program main

    use omp_lib 
    implicit none
!    !$omp requires unified_shared_memory
    ! Size of vectors
    integer :: n = 10000000
 
    ! Input vectors and Output vector
    real(8),dimension(:),allocatable :: a, b, c
    integer :: i
    real(8) :: sum
    real(8) :: startt, endt

    ! Allocate memory for each vector
    allocate(a(n), b(n), c(n))
 
    ! Initialize input vectors.
    do i=1,n
        a(i) = sin(dble(i)*1.0d0)*sin(dble(i)*1.0d0)
        b(i) = cos(dble(i)*1.0d0)*cos(dble(i)*1.0d0) 
        c(i) = 0.0d0
    enddo

    !meassure after warmup kernel
    startt=omp_get_wtime()
    ! Sum each component of arrays

    do i=1,n
        c(i) = a(i) + b(i)
    enddo

    ! Sum up vector c. Print result divided by n. It should equal 1
    sum = 0.0d0
    do i=1,n
        sum = sum +  c(i)
    enddo

    sum = sum/dble(n)
    write(*,'("Final result: ",f10.6)') sum

    endt=omp_get_wtime()
    write(*,'("Runtime is: ",f8.6," secs")') endt-startt
 
    ! Deallocate memory
    deallocate(a, b, c)
 
end program

File: Pragma_Examples/OpenMP/C/3_vecadd/4_vecadd_async_usm/vecadd.c
/*
Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.

*/
#include <stdio.h>
//#include <stdlib.h>
#include <omp.h>
#include <math.h>


#pragma omp requires unified_shared_memory

int main()
{
  int N = 100000;

  // Input arrays
  double a[N];
  double b[N];
  // Output arrays
  double c[N];

  double sum;

  #pragma omp target teams distribute parallel for nowait depend(out:a)
  for (int i=0; i<N; i++){
    a[i]=sin((double)(i+1))*sin((double)(i+1));
  }
  #pragma omp target teams distribute parallel for nowait depend(out:b)
  for (int i=0; i<N; i++){
    b[i]=cos((double)(i+1))*cos((double)(i+1));
  }
  #pragma omp target teams distribute parallel for nowait depend(out:c)
  for (int i=0; i<N; i++){
    c[i]=0.0;
  }
  double tstart = omp_get_wtime();
  #pragma omp target teams distribute parallel for nowait depend(in:a,b) depend(inout:c)
  for (int j = 0; j< N; j++){
    c[j] = a[j] + b[j];
  }

  sum = 0.0;
  #pragma omp target teams distribute parallel for reduction(+: sum) nowait depend(in:c)
  for (int i=0; i<N; i++){
    sum += c[i];
  }
  #pragma omp taskwait

  sum = sum/(double)N;
  printf("Runtime is: %lf secs\n", omp_get_wtime()-tstart);

  printf("Final result: %lf\n",sum);

}
